File: RTM Learner/RTM Learner/Pipeline/Translator.swift
import Foundation

struct Translator {
    private static let promptTemplate = """
    You are a Chinese-to-German language expert helping a German speaker learn Mandarin.

    For each word/idiom in the JSON array below, add two fields:
    - "german": a concise, natural German definition
    - "example_de": a natural German translation of the Chinese example sentence

    The words appear in a text about: {topic}
    Use this context to choose the most fitting German meaning where ambiguous.

    Input JSON array:
    {words}

    Return ONLY a JSON array (same length, same order) with the added fields.
    No markdown, no explanation.
    """

    /// Adds `german` and `example_de` to all words and idioms in the episode in-place.
    static func translate(episode: inout Episode, provider: LLMProvider) async throws {
        let all = episode.words + episode.idioms
        guard !all.isEmpty else { return }

        let payload = all.map { w in
            ["chinese": w.chinese, "pinyin": w.pinyin, "english": w.english,
             "example_zh": w.exampleZh, "example_en": w.exampleEn]
        }
        let payloadData = try JSONSerialization.data(withJSONObject: payload, options: .prettyPrinted)
        let payloadString = String(data: payloadData, encoding: .utf8) ?? "[]"

        let prompt = promptTemplate
            .replacingOccurrences(of: "{topic}", with: episode.title)
            .replacingOccurrences(of: "{words}", with: payloadString)

        let raw = try await provider.complete(prompt: prompt)
        guard let data = JSONRepair.repair(raw) else { return }

        struct TranslationResult: Decodable {
            let german: String?
            let exampleDe: String?
            enum CodingKeys: String, CodingKey {
                case german; case exampleDe = "example_de"
            }
        }
        guard let results = try? JSONDecoder().decode([TranslationResult].self, from: data) else { return }

        let n = episode.words.count
        for (i, result) in results.prefix(all.count).enumerated() {
            if i < n {
                episode.words[i].german    = result.german    ?? ""
                episode.words[i].exampleDe = result.exampleDe ?? ""
            } else {
                episode.idioms[i - n].german    = result.german    ?? ""
                episode.idioms[i - n].exampleDe = result.exampleDe ?? ""
            }
        }
    }
}

File: python/translator.py
"""Translate word lists from Chinese/English to German using the configured LLM provider."""
import json
from dataclasses import replace

from json_repair import repair_json

from domain.models import VocabEntry
from llm import call_llm, clean_json

_PROMPT = """
You are a Chinese-to-German language expert helping a German speaker learn Mandarin.

For each word/idiom in the JSON array below, add two fields:
- "german": a concise, natural German definition (not just a literal translation of the English)
- "example_de": a natural German translation of the provided Chinese example sentence

The words appear in a text about: {topic}
Use this context to choose the most fitting German meaning where the Chinese word is ambiguous.

Input JSON array:
{words}

Return ONLY a JSON array (same length, same order) with the added "german" and "example_de" fields.
No markdown, no explanation.
"""


def translate_words(
    words: list[VocabEntry],
    idioms: list[VocabEntry],
    topic: str,
) -> tuple[list[VocabEntry], list[VocabEntry]]:
    """
    Return new VocabEntry lists with 'german' and 'example_de' fields added.
    Inputs are not modified.
    """
    all_items = words + idioms
    if not all_items:
        return words, idioms

    payload = [
        {
            "chinese": w.chinese,
            "pinyin": w.pinyin,
            "english": w.english,
            "example_zh": w.example_zh,
            "example_en": w.example_en,
        }
        for w in all_items
    ]

    prompt = _PROMPT.format(topic=topic, words=json.dumps(payload, ensure_ascii=False, indent=2))
    raw = call_llm(prompt, max_tokens=4096)
    cleaned = clean_json(raw)
    try:
        translated = json.loads(cleaned)
    except json.JSONDecodeError:
        translated = json.loads(repair_json(cleaned))

    enriched = [
        replace(item, german=t.get("german", ""), example_de=t.get("example_de", ""))
        for item, t in zip(all_items, translated)
    ]

    n = len(words)
    return enriched[:n], enriched[n:]
